bpf/bpf_netdev.c

```diff
@@ -241,8 +240,17 @@ static inline int handle_ipv6(struct __sk_buff *skb, __u32 src_identity)
 	dst = (union v6addr *) &ip6->daddr;
 	info = ipcache_lookup6(&IPCACHE_MAP, dst, V6_CACHE_KEY_LEN);
 	if (info != NULL && info->tunnel_endpoint != 0) {
-		return encap_and_redirect_with_nodeid(skb, info->tunnel_endpoint,
-						      secctx, TRACE_PAYLOAD_LEN);
+		int ret = encap_and_redirect_with_nodeid(skb, info->tunnel_endpoint,
+							 secctx, TRACE_PAYLOAD_LEN);
+
+		/* If IPSEC is needed recirc through ingress to use xfrm stack
+		 * and then result will routed back through bpf_netdev on egress
+		 * but with encrypt marks.
+		 */
+		if (ret == IPSEC_ENDPOINT)
+			return redirect(HOST_IFINDEX, BPF_F_INGRESS);
```

Comment:
This corresponds to the following line from the commit message right?

> skb needs ipsec, sent through veth pair causing ingress program of cilium_net (bpf_ipsec) to run.

The main point here being that we redirect to ingress on that device so that we don't lose the existing `skb->cb[]` fields, and hence everything will work OK.

Reply:
yeah that corresponds to the comment. But actually I'm going to convert that to 'return TC_ACT_OK' actually I already did it for ipv4.

The theory is we let the veth (cilium_host) pair forward it as normal, this calls dev_skb_forward into cilium_net. Here cb[] is in tact from above. Then bpf_ipsec can translate cb[] into mark values. Its really just a trick to revert clearing of the mark field done by dev_skb_forward between the veth pairs.

Reply:
Ah, I kept thinking that it would be cleared when you cross the veth pair, but I suppose that's primarily the case when you're also crossing a netns boundary, not so much when both ends of the veth are in the init netns.

Reply:
The packet is "scrubbed" when you cross the veth pair so that mark and such are cleared but the cb[] is not so we can use like above. I don't believe cb[] is cleared upon crossing netns boundaries either which would be handled by the dev_skb_forward calling the scrub routine same as is done with veth pair in same netns.